pkg/cmd/pipeline/start.go

```diff
@@ -92,44 +109,210 @@ tkn pipeline start foo --param NAME=VALUE --resource source=scaffold-git  -s Ser
 			return NameArg(args, p)
 		},
 		RunE: func(cmd *cobra.Command, args []string) error {
-
-			opt := startOptions{
-				Resources:          res,
-				Params:             params,
-				ServiceAccountName: svc,
-				ServiceAccounts:    svcs,
-				Last:               last,
+			opt.stream = &cli.Stream{
+				Out: cmd.OutOrStdout(),
+				Err: cmd.OutOrStderr(),
 			}
-			return startPipeline(cmd.OutOrStdout(), opt, p, args[0])
+			return opt.run(args[0])
 		},
 	}
 
-	c.Flags().StringSliceVarP(&res, "resource", "r", []string{}, "pass the resource name and ref")
-	c.Flags().StringSliceVarP(&params, "param", "p", []string{}, "pass the param")
-	c.Flags().StringVarP(&svc, "serviceaccount", "s", svc, "pass the serviceaccount name")
-	c.Flags().StringSliceVar(&svcs, "task-serviceaccount", []string{}, "pass the service account corresponding to the task")
-	c.Flags().BoolVarP(&last, "last", "l", false, "re-run the pipeline using last pipelinerun values")
+	//c.Flags().StringSliceVarP(&opt.Resources, "resource", "r", []string{}, "pass the resource name and ref")
+	//c.Flags().StringSliceVarP(&opt.Params, "param", "p", []string{}, "pass the param")
+	c.Flags().StringVarP(&opt.ServiceAccountName, "serviceaccount", "s", "", "pass the serviceaccount name")
+	c.Flags().StringSliceVar(&opt.ServiceAccounts, "task-serviceaccount", []string{}, "pass the service account corresponding to the task")
+	c.Flags().BoolVarP(&opt.Last, "last", "l", false, "re-run the pipeline using last pipelinerun values")
 	return c
 }
 
-func startPipeline(out io.Writer, opt startOptions, p cli.Params, pName string) error {
+func (opt *startOptions) run(pName string) error {
+	if !opt.Last {
+		if err := opt.getInput(pName); err != nil {
+			return err
+		}
+	}
+
+	return opt.startPipeline(pName)
+}
+
+func (opt *startOptions) getInput(pname string) error {
+	cs, err := opt.params.Clients()
+	if err != nil {
+		return err
+	}
+	pres, err := getPipelineResources(cs.Tekton, opt.params.Namespace())
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to list pipelineresources from %s namespace \n", opt.params.Namespace())
+		return err
+	}
+
+	resources := getPipelineResourcesByFormat(pres.Items)
+
+	pipeline, err := getPipeline(cs.Tekton, opt.params.Namespace(), pname)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to get pipeline %s from %s namespace \n", pname, opt.params.Namespace())
+		return err
+	}
+
+	err = opt.getInputResources(resources, pipeline)
+	if err != nil {
+		return err
+	}
+
+	err = opt.getInputParams(pipeline)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (opt *startOptions) getInputResources(resources []resourceOptions, pipeline *v1alpha1.Pipeline) error {
+	var ans string
+	for _, res := range pipeline.Spec.Resources {
+		options := getOptionsByFilter(resources, string(res.Type))
```

Comment:
later on its better to avoid the repeated computations

Reply:
you want me to filter the resource by type also upfront and store all different ?

Reply:
@piyush-garg yes it is okay for now :smile: . Will fix it if we face any issues down the line

Reply:
I have addressed this and updated the PR